File: HoopsCam-iOS/HoopsCAM/Models/BBoxActivityDetector.swift
import Foundation
import CoreGraphics

class BBoxActivityDetector {
    // Configuration
    private let historySize = 60 // 2 seconds at 30fps
    private let shapeThreshold: Double = 0.002 // Hand-tuned threshold from Python
    
    // State: Map of TrackerID -> [BBoxFeatures]
    private var history: [UUID: [BBoxFeatures]] = [:]
    
    struct BBoxFeatures {
        let width: Double
        let height: Double
        let area: Double
        let aspectRatio: Double
    }
    
    func isActive(trackerId: UUID, bbox: CGRect) -> Bool {
        let features = extractFeatures(bbox)
        
        // Update history
        if history[trackerId] == nil {
            history[trackerId] = []
        }
        history[trackerId]?.append(features)
        
        // Maintain buffer size
        if let count = history[trackerId]?.count, count > historySize {
            history[trackerId]?.removeFirst(count - historySize)
        }
        
        // Check activity
        let variance = calculateShapeVariance(trackerId: trackerId)
        return variance > shapeThreshold
    }
    
    private func extractFeatures(_ bbox: CGRect) -> BBoxFeatures {
        // Assume bbox is normalized [0-1]
        let w = Double(bbox.width)
        let h = Double(bbox.height)
        return BBoxFeatures(
            width: w,
            height: h,
            area: w * h,
            aspectRatio: h > 0 ? w / h : 0
        )
    }
    
    private func calculateShapeVariance(trackerId: UUID) -> Double {
        guard let trackHistory = history[trackerId], trackHistory.count > 10 else {
            return 1.0 // Assume active if not enough history
        }
        
        // Calculate variance of Aspect Ratio and Area
        let aspectRatios = trackHistory.map { $0.aspectRatio }
        let areas = trackHistory.map { $0.area }
        
        let aspectVar = variance(aspectRatios)
        let areaVar = variance(areas)
        
        // Combine them (simple average or weighted)
        // Python: (aspect_std + area_std) / 2 ... wait, std or var?
        // Python code said: (std(aspect) + std(area)) / 2
        // Let's implement Standard Deviation
        
        return (sqrt(aspectVar) + sqrt(areaVar)) / 2.0
    }
    
    private func variance(_ numbers: [Double]) -> Double {
        let count = Double(numbers.count)
        if count == 0 { return 0 }
        
        let mean = numbers.reduce(0, +) / count
        let sumSquaredDiffs = numbers.reduce(0) { sum, val in
            sum + pow(val - mean, 2)
        }
        return sumSquaredDiffs / count
    }
    
    func clear() {
        history.removeAll()
    }
}

File: app/src/main/java/com/sportscam/domain/BBoxActivityDetector.kt
package com.sportscam.domain

import android.graphics.RectF
import com.sportscam.data.models.Track
import kotlin.math.abs

class BBoxActivityDetector(
    private var shapeThreshold: Float = 0.002f, // Default matching iOS
    private val historySize: Int = 60           // 2 seconds at 30fps
) {
    private val minHistoryForDecision: Int = 10
    
    private data class BBoxFeatures(
        val width: Float,
        val height: Float,
        val area: Float,
        val aspectRatio: Float
    )

    // Map of Track ID -> List of features
    private val history = HashMap<Int, MutableList<BBoxFeatures>>()

    fun updateThreshold(threshold: Float) {
        this.shapeThreshold = threshold
    }

    /**
     * Determines if a track is "Active" (moving) based on bbox shape variance.
     */
    fun isActive(track: Track): Boolean {
        val features = extractFeatures(track.bbox)
        
        val trackHistory = history.getOrPut(track.id) { ArrayList() }
        trackHistory.add(features)
        
        if (trackHistory.size > historySize) {
            trackHistory.removeAt(0)
        }
        
        // Calculate activity score
        val varianceScore = calculateShapeVariance(track.id)
        return varianceScore > shapeThreshold
    }
    
    private fun extractFeatures(bbox: RectF): BBoxFeatures {
        val w = bbox.width()
        val h = bbox.height()
        val safeH = maxOf(h, 0.0001f)
        
        return BBoxFeatures(
            width = w,
            height = h,
            area = w * h,
            aspectRatio = w / safeH
        )
    }

    private fun calculateShapeVariance(trackId: Int): Float {
        val trackHistory = history[trackId] ?: return 1.0f // Assume active if no history
        
        if (trackHistory.size < minHistoryForDecision) {
            return 1.0f // Assume active initially
        }
        
        val aspectRatios = trackHistory.map { it.aspectRatio }
        val areas = trackHistory.map { it.area }
        
        val aspectVar = calculateVariance(aspectRatios)
        val areaVar = calculateVariance(areas)
        
        // Combine: (std(aspectRatio) + std(area)) / 2
        return (kotlin.math.sqrt(aspectVar) + kotlin.math.sqrt(areaVar)) / 2.0f
    }

    private fun calculateVariance(numbers: List<Float>): Float {
        if (numbers.isEmpty()) return 0f
        val mean = numbers.average().toFloat()
        return numbers.map { (it - mean) * (it - mean) }.average().toFloat()
    }

    fun cleanup(activeTrackIds: Set<Int>) {
        val iterator = history.entries.iterator()
        while (iterator.hasNext()) {
            if (!activeTrackIds.contains(iterator.next().key)) {
                iterator.remove()
            }
        }
    }
    
    fun reset() {
        history.clear()
    }
}
